### PR TITLE
Continue #360

### PR DESCRIPTION
**Pull request by iam-py-test**
## Summary
This continues the discussion on #360 

## Checklist

- [ ] I have verified that I have not modified any files inside the `alt-version` folder (automated code will automatically update those files)

- [ ] I have verified that I have not modified any files inside the `dnsmasq-version` folder (automated code will automatically update those files)

### CODE REVIEW

**Review by blocklistproject — COMMENTED** [review]
@fishcharlie if this is merged, don't we just have to include line 16 in the automation yml?

**Comment by iam-py-test** [comment]
@blocklistproject can you resolve the conflicts

**Comment by iam-py-test** [comment]
@fishcharlie would you mind looking over check-dns.js
There is a mysterious error I can't fix and you seem more familiar with Node.js than I am.

**Comment by fishcharlie** [comment]
@iam-py-test Yep for sure! Might not be until this weekend tho. Pretty busy with personal stuff.

**Comment by iam-py-test** [comment]
> 
> 
> @iam-py-test Yep for sure! Might not be until this weekend tho. Pretty busy with personal stuff.

Thank you!
I will be busy over this weekend too, so I can understand.

**Comment by fishcharlie** [comment]
@iam-py-test No problem. I'll review it when I can and then give feedback and you can look it over when you get a chance. No rush at all.

**Comment by iam-py-test** [comment]
@fishcharlie I have an idea for detecting parked domains; if we could find a list of known domain-parking-related-domains, we can then test if the domains redirect to one of those.

**Comment by fishcharlie** [comment]
@iam-py-test Ahhh. Super clever idea! Love it. However I think that should follow the same system of adding to some type of document to be reviewed manually.

Why don't we create a separate issue for that idea? Wanna focus on one thing at a time so it doesn't get too big too quickly to the point where we can't manage it anymore.

**Comment by iam-py-test** [comment]
> @iam-py-test Ahhh. Super clever idea! Love it. However I think that should follow the same system of adding to some type of document to be reviewed manually.
> 
> Why don't we create a separate issue for that idea? Wanna focus on one thing at a time so it doesn't get too big too quickly to the point where we can't manage it anymore.

Ok. We should complete this, and one this is done, start on that.
Maybe we should create a meta issue for this, or is the discussion ok?

**Comment by fishcharlie** [comment]
Working on this now.

Will provide an update soon.

**Comment by fishcharlie** [comment]
I'm closing this PR.

This PR is not thought through **at all**. Here are the issues I'm running into:

First of all, I don't have write access to your branch. So the errors you wanted me to fix, I can't push the changes. I have left the modified file code below.

Second, there are merge conflicts. Did you edit the everything list? The only thing this PR should add is a new file to run the check. It shouldn't be modifying any lists or anything.

Finally (and the primary reason I'm closing this PR). There are almost 4 million entries in the lists. That is a **TON** of entries. I used the `1.1.1.1` DNS server to make these queries, and it took over 10 minutes to complete. Along with that, some queries failed even tho there is an DNS record for the domain. I'm assuming this is due to some type of DDOS protection to avoid making almost 4 million requests all at once. That doesn't even take into account running this on GitHub Actions where they probably have additional security measures in place to prevent this type of behavior.

Running this script resulted in over 3 million invalid entries. Again, probably due to server abuse prevention measures in place. That is a **LOT** more than I expected. However, even with say 10k (totally guessing on that number) invalid entries, that is still a **LOT** of effort to put into verification to ensure it really should be removed.

I'm still **very** against having this script automatically remove all invalid entires. For example, one of the last entries that showed up when running the script was `familyporn.tv`. That **clearly** should exist on the list no matter if the DNS entry doesn't exist anymore. And just to prove my point about DNS server abuse protection, when querying for the DNS entry again I got a valid DNS entry: `familyporn.tv.		60	IN	A	51.79.19.103`.

However, there are some perfectly valid things that should be removed. Such as `a37ubfu6nmjxzcff.onion`. So it did accurately capture some benefits. Is running a DNS query script worth it to remove domains like that? No way. It'd be a lot better to go through and write a script to remove all domains ending in `.onion`. Or move them to a different list (not sure if TOR has some type of blocking system).

Also, what happens if the DNS server is down or offline for some reason? No logic is in place for that.

This script is very abusive to a DNS server and therefore will not be accepted.

Now, could we do something where we limit the number of DNS requests made per minute? Say maybe 100 per second? That would mean the script would take over 11 hours to run. And I'm sure GitHub Actions has a timeout at a certain point. I still think 100 DNS requests per second for 11+ hours straight might be considered abusive behavior towards a DNS server, but that is besides the point.

That isn't even to mention my thoughts about how useless this PR is in the first place. I **STILL** don't understand the purpose of it. What problem are we trying to solve here? No one has given a clear and concise answer to that question that makes it worth it.

One thing I should mention. I'm well aware that it's not the most efficient. We could delay each DNS request to be sequential and wait for a reply before making the next DNS query. That would be pretty easy to do by converting it to use for loops instead of `forEach`. But I'm not going to put the time into that because still making 4 million DNS requests back to back, still seems very abusive. And all the issues I mentioned above about amount of time it'd take and all of that are still valid.

I think that covers most of it.

---

@iam-py-test I don't think a PR like this will ever be accepted in the foreseeable future. However, if you would like to try to continue this and have ideas about how to address the things I mentioned above, please create an issue with a **detailed** proposal.

This detailed proposal should include:

- A clear explanation of why this is necessary. Explain the benefits and why you believe the benefits are greater than the drawbacks (time, effort, etc).
- How to address the DNS server abuse issue
- How to address temporary glitches (DNS server down, network temporarily down, etc)
- How to address the time it takes to run this script (**especially** if we run a rate limiter, as a solution for the DNS server abuse issue)
- Explanation of why we should be putting effort into this, as opposed to other ideas to reduce the amount of time it takes to correct false positives, or making it easier to report false positives.
- How to manage the process of manually checking each domain to ensure it's valid or not and ensure that is done in a timely manner

There are likely more things that this proposal would need to include. But the general idea is to have it so bulletproof and detailed that no issues can arise from it. This proposal should make absolutely crystal clear that this is a clear, beneficial, and major win for the project, with detail about how that will be achieved.

---

Tldr.

I'm closing this PR since it has a lot of architectural issues that have not been considered or thought through at all. A detailed proposal describing how to fix these issues **might** cause this to be reconsidered, but until then, I don't see this being approved anytime in the foreseeable future.

---

**check-dns.js**

```js
const dns = require("dns").promises;
const fs = require("fs").promises;
const path = require("path");
const lists = require("./utils/lists");

dns.setServers(["1.1.1.1"]);

(async () => {
	console.log(Date.now());
	const invalidDomains = new Set();

	// For each non beta list
	await Promise.all(lists.map(async (list) => {
		// Read list contents
		const listContents = await fs.readFile(path.join(__dirname, "..", `${list}.txt`), "utf8");
		// Get array of strings for each line of the list contents
		const lines = listContents.split("\n");
		await Promise.all(lines.map(async (line) => {
			// If line starts with "0.0.0.0 " (not "#" or empty lines)
			if (line.startsWith("0.0.0.0 ")) {
				// Remove "0.0.0.0 " & comments after domain ("#") to extract domain
				const domain = line.replace("0.0.0.0 ", "").split("#")[0].trim();

				try {
					await dns.resolve(domain);
					// DNS record is valid
				} catch (error) {
					console.log(domain);
					// DNS record is invalid
					invalidDomains.add(domain);
				}
			}
		}));
	}));

	// Do something with `invalidDomains`. Likely create GitHub issues or something for each domain
	// We could also write it to a file or something

	console.log(Date.now());
})();
```

**utils/lists.js**

```js
const lists = [
	"abuse",
	"ads",
	"crypto",
	"drugs",
	"facebook",
	"fraud",
	"gambling",
	"malware",
	"phishing",
	"piracy",
	"porn",
	"ransomware",
	"redirect",
	"scam",
	"tiktok",
	"torrent",
	"tracking",

	// The following lists are in beta and therefore not included in the everything list:

	// "smart-tv",
	// "basic",
	// "whatsapp"
];

module.exports = lists;
```
